presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateEmptyJoins.java

```diff
@@ -15,21 +15,43 @@
 
 import com.facebook.presto.matching.Captures;
 import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.spi.plan.Assignments;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.ProjectNode;
 import com.facebook.presto.spi.plan.ValuesNode;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
 import com.facebook.presto.sql.planner.iterative.Rule;
 import com.facebook.presto.sql.planner.plan.JoinNode;
 
+import java.util.Collection;
 import java.util.Collections;
 
 import static com.facebook.presto.SystemSessionProperties.isEmptyJoinOptimization;
+import static com.facebook.presto.spi.plan.ProjectNode.Locality.LOCAL;
 import static com.facebook.presto.sql.planner.optimizations.QueryCardinalityUtil.isAtMost;
 import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.facebook.presto.sql.relational.Expressions.constantNull;
 
 public class EliminateEmptyJoins
         implements Rule<JoinNode>
 {
     private static final Pattern<JoinNode> PATTERN = join();
 
+    // Build assignment list for the new Project node as: X=X if X is from non-empty child and X=null otherwise.
+    public static Assignments buildAssignments(Collection<VariableReferenceExpression> variables, PlanNode nonEmptyChild)
+    {
+        Assignments.Builder builder = Assignments.builder();
+        variables.forEach(variable -> {
+            if (nonEmptyChild.getOutputVariables().contains(variable)) {
+                builder.put(variable, variable);
+            }
+            else {
+                builder.put(variable, constantNull(variable.getType()));
```

Comment:
What happens if there was a const/literal on the empty side and may be we projected it out? like:

SELECT T1.*, T2.* FROM T1 LEFT OUTER JOIN (SELECT 1 AS x, .. FROM T2 LIMIT 0) USING(key)

Good to add this as a test case.

Reply:
Good to add but the result should be null in this case too.

Reply:
Added more tests that covers this case.